### PR TITLE
feat: Adding support for Windows testing

### PR DESCRIPTION
**Pull request by owenrumney**
## Description
- replace hard coded file paths to use filepath.Join()
- add a load of deferred closure for test tempdir (Windows doesn't like
  not closing first)

## Notes
- Integration and Module Integration tests only run on Ubuntu at the moment
- Some tests not applicable for Windows are skipped with an explanation

trivy_on_windows

## Related issues
- Gets closer to achieving https://github.com/aquasecurity/trivy/issues/2058

## Related PRs
- [ ] feeds into #2592

### CODE REVIEW

File: pkg/fanal/analyzer/pkg/apk/apk_test.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 )
 
 func TestParseApkInfo(t *testing.T) {
+	t.Skip("Not relevant for Windows testing")
```
Comment (knqyf263) [inline pkg/fanal/analyzer/pkg/apk/apk_test.go:14]: It always seems to get skipped. Am I missing something? 
Also, this package is essential for scanning Alpine Linux. I think we need to run the test even on Windows.

File: pkg/fanal/analyzer/secret/secret_test.go [hunk 2]
```diff
@@ -14,6 +15,8 @@ import (
 )
 
 func TestSecretAnalyzer(t *testing.T) {
+	t.Skip("skipping test in short mode.")
```
Comment (knqyf263) [inline pkg/fanal/analyzer/secret/secret_test.go:18]: Don't we need `testing.Short()`?

```suggestion
	if testing.Short() {
		t.Skip("skipping test in short mode.")
	}
```

File: pkg/fanal/artifact/remote/git_test.go [hunk 3]
```diff
@@ -1,3 +1,6 @@
+//go:build linux || bsd || darwin || freebsd
```
Comment (knqyf263) [inline pkg/fanal/artifact/remote/git_test.go:1]: We might be able to leverage `unix`.

```suggestion
//go:build unix
```
https://tip.golang.org/doc/go1.19#go-unix

File: pkg/fanal/handler/misconf/misconf_test.go [hunk 4]
```diff
@@ -80,44 +87,158 @@ func Test_FindingFSTarget(t *testing.T) {
...
+			input:      []string{string(os.PathSeparator), filepath.Join(string(os.PathSeparator), "bar")},
+			wantTarget: string(os.PathSeparator),
 			wantPaths:  []string{".", "bar"},
 		},
 	}
 
 	for _, test := range tests {
 		t.Run(fmt.Sprintf("%#v", test.input), func(t *testing.T) {
+
+			if runtime.GOOS == "windows" {
```
Comment (knqyf263) [inline pkg/fanal/handler/misconf/misconf_test.go:140]: Seems like it already skips Windows. Do we need these lines?
https://github.com/aquasecurity/trivy/blob/0c056195650e38aa166c360412ebfe1b013f4510/pkg/fanal/handler/misconf/misconf_test.go#L75-L77

File: pkg/fanal/handler/misconf/misconf_test.go [hunk 5]
```diff
@@ -80,44 +87,158 @@ func Test_FindingFSTarget(t *testing.T) {
...
+			input:      []string{string(os.PathSeparator), filepath.Join(string(os.PathSeparator), "bar")},
+			wantTarget: "",
+			wantPaths:  []string{"/", "/bar"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%#v", test.input), func(t *testing.T) {
+
+			if runtime.GOOS == "windows" {
```
Comment (knqyf263) [inline pkg/fanal/handler/misconf/misconf_test.go:230]: Do we need the if-statement as only windows reaches here?
https://github.com/aquasecurity/trivy/blob/0c056195650e38aa166c360412ebfe1b013f4510/pkg/fanal/handler/misconf/misconf_test.go#L165-L167

File: pkg/fanal/image/daemon/podman_test.go [hunk 6]
```diff
@@ -43,6 +43,7 @@ func setupPodmanSock(t *testing.T) *httptest.Server {
 }
 
 func TestPodmanImage(t *testing.T) {
+	t.Skip("podman.sock is not available for Windows CI")
```
Comment (knqyf263) [inline pkg/fanal/image/daemon/podman_test.go:46]: Looks like it skips anytime.

```suggestion
	if runtime.GOOS == "windows" {
		t.Skip("podman.sock is not available for Windows CI")
	}
```

File: pkg/fanal/walker/tar_test.go [hunk 7]
```diff
@@ -16,6 +17,9 @@ import (
 )
 
 func TestLayerTar_Walk(t *testing.T) {
+	if runtime.GOOS == "windows" {
```
Comment (knqyf263) [inline pkg/fanal/walker/tar_test.go:20]: This test is essential for container image scanning. Is it hard to run tests on Windows?

File: pkg/licensing/classifier.go [hunk 8]
```diff
@@ -96,13 +97,26 @@ func googleClassifierLicense(filePath string, contents []byte) types.LicenseFile
...
+
+		if _, ok := findingsMap[finding]; !ok {
+			findingsMap[finding] = true
+		}
 	}
 
+	var findings []types.LicenseFinding
+	for finding := range findingsMap {
+		findings = append(findings, finding)
+	}
```
Comment (knqyf263) [inline pkg/licensing/classifier.go:114]: What if using maps.Keys?
https://pkg.go.dev/golang.org/x/exp/maps#Keys

File: pkg/licensing/classifier.go [hunk 9]
```diff
@@ -96,13 +97,26 @@ func googleClassifierLicense(filePath string, contents []byte) types.LicenseFile
...
+		finding := types.LicenseFinding{
 			Name:       m.Name,
 			Confidence: m.Confidence,
 			Link:       licenseLink,
-		})
+		}
+
+		if _, ok := findingsMap[finding]; !ok {
+			findingsMap[finding] = true
+		}
```
Comment (knqyf263) [inline pkg/licensing/classifier.go:108]: I'd like to know the reason we should check the key existence. Doesn't the following work?

```suggestion
		findingsMap[finding] = true
```

File: pkg/module/module_test.go [hunk 10]
```diff
@@ -16,6 +16,7 @@ import (
 )
 
 func TestManager_Register(t *testing.T) {
+	t.Skip("Test satisfied adequately by Linux tests")
```
Comment (knqyf263) [inline pkg/module/module_test.go:19]: Please correct me if I'm wrong.

```suggestion
	if runtime.GOOS == "windows" {
		t.Skip("Test satisfied adequately by Linux tests")
	}
```

File: pkg/plugin/plugin_test.go [hunk 11]
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 func TestPlugin_Run(t *testing.T) {
+	t.Skip("Test satisfied adequately by Linux tests")
```
Comment (knqyf263) [inline pkg/plugin/plugin_test.go:17]: ditto

File: pkg/cloud/aws/commands/run_test.go [hunk 12]
```diff
@@ -628,6 +630,9 @@ deny[res] {
 			}
 
 			if test.regoPolicy != "" {
+				if runtime.GOOS == "windows" {
+					t.Skip("Skipping test on windows")
+				}
```
[the remark this answers is not in the document]
Comment (owenrumney) [reply pkg/cloud/aws/commands/run_test.go:635]: It didn't but I've found the issue in defsec so its resolved.

File: pkg/module/module_test.go (hunk 10 above)
Comment (owenrumney) [reply pkg/module/module_test.go]: fixed

File: pkg/plugin/plugin_test.go (hunk 11 above)
Comment (owenrumney) [reply pkg/plugin/plugin_test.go]: fixed in all the places

File: pkg/licensing/classifier.go (hunk 9 above)
Comment (owenrumney) [reply pkg/licensing/classifier.go]: no reason, it works fine. changed

File: pkg/licensing/classifier.go (hunk 8 above)
Comment (owenrumney) [reply pkg/licensing/classifier.go]: yup, good good

File: pkg/fanal/analyzer/pkg/apk/apk_test.go (hunk 1 above)
Comment (owenrumney) [reply pkg/fanal/analyzer/pkg/apk/apk_test.go]: this is passing the tests now

File: pkg/fanal/analyzer/analyzer_test.go [hunk 13]
```diff
@@ -331,12 +337,13 @@ func TestAnalyzeFile(t *testing.T) {
 					"lib/ld-musl-x86_64.so.1",
 				},
 			},
+			skipWindows: true,
```
[the remark this answers is not in the document]
Comment (owenrumney) [reply pkg/fanal/analyzer/analyzer_test.go]: Fixed this - removed the requirement for `skipWindows`

File: .github/workflows/test_macos.yaml [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+      - 'mkdocs.yml'
+      - 'LICENSE'
+  pull_request:
+env:
+  GO_VERSION: "1.19"
+  TINYGO_VERSION: "0.25.0"
+jobs:
+  test:
+    name: Test - macOS
+    runs-on: macos-latest
```
[the remark this answers is not in the document]
Comment (owenrumney) [reply .github/workflows/test_macos.yaml:18]: I'll have another look at this. I did it in a single file initially, but it was very ugly with so many conditions

[the remark this answers is not in the document]
Comment (owenrumney) [reply .github/workflows/test_macos.yaml]: Okay - done, not as ugly as I was worried it would be

**Comment by prabhu** [comment]
@owenrumney Like many, I can't wait to have this support. In the interim, is it possible to release a small PR that only does `replace hard coded file paths to use filepath.Join()`?

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**